linalg/sparsemat.cpp

```diff
@@ -469,42 +469,34 @@ void SparseMatrix::SortColumnIndices()
 #if defined(MFEM_USE_CUDA)
       size_t pBufferSizeInBytes = 0;
       void *pBuffer = NULL;
+      int *P = NULL;
 
       const int n = Height();
       const int m = Width();
       const int nnzA = J.Capacity();
       real_t * d_a_sorted = ReadWriteData();
       const int * d_ia = ReadI();
       int * d_ja_sorted = ReadWriteJ();
-      csru2csrInfo_t sortInfoA;
 
       cusparseMatDescr_t matA_descr;
       cusparseCreateMatDescr( &matA_descr );
       cusparseSetMatIndexBase( matA_descr, CUSPARSE_INDEX_BASE_ZERO );
       cusparseSetMatType( matA_descr, CUSPARSE_MATRIX_TYPE_GENERAL );
 
-      cusparseCreateCsru2csrInfo( &sortInfoA );
-
-#ifdef MFEM_USE_SINGLE
-      cusparseScsru2csr_bufferSizeExt( handle, n, m, nnzA, d_a_sorted, d_ia,
-                                       d_ja_sorted, sortInfoA,
-                                       &pBufferSizeInBytes);
-#elif defined MFEM_USE_DOUBLE
-      cusparseDcsru2csr_bufferSizeExt( handle, n, m, nnzA, d_a_sorted, d_ia,
-                                       d_ja_sorted, sortInfoA,
-                                       &pBufferSizeInBytes);
-#else
-      MFEM_ABORT("Floating point type undefined");
-#endif
+      cusparseXcsrsort_bufferSizeExt(handle, n, m, nnzA, d_ia, d_ja_sorted,
+                                     &pBufferSizeInBytes);
 
       CuMemAlloc( &pBuffer, pBufferSizeInBytes );
 
+      cusparseCreateIdentityPermutation(handle, nnzA, P);
+      cusparseXcsrsort(handle, n, m, nnzA, matA_descr, d_ia, d_ja_sorted, P, pBuffer);
+
 #ifdef MFEM_USE_SINGLE
-      cusparseScsru2csr( handle, n, m, nnzA, matA_descr, d_a_sorted, d_ia,
-                         d_ja_sorted, sortInfoA, pBuffer);
+      cusparseSgthr(handle, nnzA, d_a_sorted, d_a_sorted, P,
+                    CUSPARSE_INDEX_BASE_ZERO);
 #elif defined MFEM_USE_DOUBLE
-      cusparseDcsru2csr( handle, n, m, nnzA, matA_descr, d_a_sorted, d_ia,
-                         d_ja_sorted, sortInfoA, pBuffer);
+      cusparseDgthr(handle, nnzA, d_a_sorted, d_a_sorted, P,
```

Comment:
With CUDA 12 this is not available any more (I am trying on CUDA 12.4). Looks like the generic `cusparseGather` should be used instead.

Reply:
> With CUDA 12 this is not available any more (I am trying on CUDA 12.4). Looks like the generic `cusparseGather` should be used instead.

I looked cusparse.h, there is no cusparseDgthr. However, in section 7.9.12, cuspareXcsrsort, it still uses cusparseDgthr.

Reply:
@pazner , could you please fix this with CUDA 12? I do not know how to do that with cusparseGather. And maybe we should also process for different CUDA versions, e.g. CUDA 11 and 12.

Reply:
Can you take a look at the [cusparseXcsrsort-update](https://github.com/mfem/mfem/tree/cusparseXcsrsort-update) branch? If that looks good, please merge it here.